File: fashionist/routes/clothes.py
from typing import Any, List, Optional, Dict

from fastapi import APIRouter, Response, HTTPException
from fastapi import UploadFile, Body

from ..models import PyObjectId
from ..models.clothes import Clothes, UpdateClothes
from ..models.wardrobe import Wardrobe

router = APIRouter(prefix='/clothes')


@router.get('/', status_code=200, response_model=List[Clothes])
async def get_all_clothes(
        tags: Optional[List[str]] = None,
        wardrobe_id: Optional[str] = None,
        offset: int = 0,
        limit: int = 100
):
    filters = {"$and": []}
    if wardrobe_id:
        wardrobe = (await Wardrobe.read(kwargs={"_id": wardrobe_id}))[0]
        filters["$and"].append({"_id": {"$in": wardrobe.clothes}})

    if tags:
        filters["$and"].append({"tags": tags})

    if not filters["$and"]:
        return await Clothes.read(offset=offset, limit=limit)

    return await Clothes.read(
        offset=offset,
        limit=limit,
        kwargs=filters
    )


@router.get('/{_id}', status_code=200, response_model=Clothes)
async def get_clothes(_id: str, wardrobe_id: Optional[str] = None):
    _id = PyObjectId(_id)

    if wardrobe_id:
        wardrobe = (await Wardrobe.read(kwargs={"_id": PyObjectId(wardrobe_id)}))[0]
        filters = {"$and": [{"_id": {"$in": wardrobe.clothes}}, {"_id": _id}]}
    else:
        filters = {"_id": _id}

    clothes = await Clothes.read(kwargs=filters)
    if not clothes:
        raise HTTPException(status_code=404, detail="Clothes not found")

    return clothes[0]


@router.post('/', status_code=201, response_model=Clothes)
async def create_clothes(value: UploadFile, tags: List[str]=Body()):
    clothes_id = await Clothes.create(Clothes(
        value=value.file.read(), 
        tags=tags, url=f'http://192.168.1.121:8080/{value.filename}'
        ))

    clothes = (await Clothes.read(kwargs=dict(_id=clothes_id)))[0]
    
    return clothes


@router.put('/{_id}', status_code=204, response_class=Response)
async def update_clothes(_id: str, clothes: UpdateClothes):
    await Clothes.update(PyObjectId(_id), clothes)


@router.delete('/{_id}', status_code=204, response_class=Response)
async def delete_clothes(_id: str, wardrobe_id: Optional[str] = None):
    _id = PyObjectId(_id)

    if wardrobe_id:
        wardrobe_id = PyObjectId(wardrobe_id)

        wardrobe = (await Wardrobe.read(kwargs={"_id": wardrobe_id}))[0]
        clothes_on_database = await Clothes.read(
            kwargs={"$and": [{"_id": {"$in": wardrobe.clothes}}, {"_id": _id}]}
        )
        if not clothes_on_database:
            return HTTPException(status_code=404, detail="Clothes not found on wardrobe")

    await Clothes.delete(_id)

File: fashionist-front/src/services/Clothes.js
import client from './Base';
import urlParamCheck from './Utils';

const endpoint = '/clothes';


export async function getAllClothes(tags, wardrobeId, offset, limit) {
    let url = endpoint;

    if (tags) {
        if (!url.endsWith('&')) {
            url += '?'
        }
        for (let tag of tags) {
            url += `tags[]=${tag}&`
        }
    }

    url = urlParamCheck(url, 'wardrobe_id', wardrobeId);
    url = urlParamCheck(url, 'offset', offset);
    url = urlParamCheck(url, 'limit', limit);


    if (url.endsWith('&')) {
        url = url.substring(0, url.length - 2);
    }

    return new Promise((resolve, reject) => {
        client.get(url)
            .then((response) => {
                resolve(response.data);
            })
            .catch((error) => {
                reject(error);
            });
    });
}

export async function getClothes(id, wardrobeId) {
    let url = `${endpoint}/${id}`;
    url = urlParamCheck(url, 'wardrobe_id', wardrobeId);

    return new Promise((resolve, reject) => {
        client.get(url)
            .then((response) => {
                resolve(response.data);
            })
            .catch((error) => {
                reject(new Error(error));
            });
    });
}

export async function createClothes(url, tags) {
    return new Promise((resolve, reject) => {
        client.post(endpoint, { url, tags })
            .then((response) => {
                resolve(response.data);
            })
            .catch((error) => {
                reject(new Error(error));
            });
    });
}

export async function updateClothes(id, tags) {
    return new Promise((resolve, reject) => {
        client.put(`endpoint/${id}`, { tags })
            .then((response) => {
                resolve(response.data);
            })
            .catch((error) => {
                reject(new Error(error));
            });
    });
}

export async function deleteCLothes(id) {
    return new Promise((resolve, reject) => {
        client.delete(`endpoint/${id}`)
            .then((response) => {
                resolve(response.data);
            })
            .catch((error) => {
                reject(new Error(error));
            });
    });
}

export default {
    getAllClothes,
    getClothes,
    createClothes,
    updateClothes,
    deleteCLothes
};